lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,502 @@
+  circt::DenseMap<Value, Value> loweredInstances;
+
+  // State to keep a necessary info when traversing potentially nested subfields
+  // of bundles: the root bundle and the fieldname suffix computed so far.
+  circt::DenseMap<Value, ValueField> loweredSubfieldInfo;
+};
+} // end anonymous namespace
+
+/// This is the pass constructor.
+std::unique_ptr<mlir::Pass> circt::firrtl::createLowerFIRRTLTypesPass() {
+  return std::make_unique<FIRRTLTypesLowering>();
+}
+
+// This is the main entrypoint for the lowering pass.
+void FIRRTLTypesLowering::runOnOperation() {
+  auto module = getOperation();
+  auto *body = module.getBodyBlock();
+
+  OpBuilder theBuilder(&getContext());
+  builder = &theBuilder;
+
+  // Lower the module block arguments.
+  SmallVector<BlockArgument, 8> args(body->args_begin(), body->args_end());
+  for (auto arg : args)
+    mlir::TypeSwitch<Type>(arg.getType())
+        .Case<BundleType>([&](BundleType type) { lowerArg(arg, type); })
+        .Default<>([&](Type type) { lowerArg(arg, type); });
+
+  // Lower the operations.
+  for (auto &op : body->getOperations()) {
+    builder->setInsertionPoint(&op);
+    if (failed(dispatchVisitor(&op)))
+      return;
+  }
+
+  cleanup();
+}
+
+//===----------------------------------------------------------------------===//
+// Lowering module block arguments.
+//===----------------------------------------------------------------------===//
+
+// Lower arguments with bundle type by flattening them.
+void FIRRTLTypesLowering::lowerArg(BlockArgument arg, BundleType type) {
+  // Flatten the bundle types.
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenBundleTypes(type, "", false, fieldTypes);
+
+  for (auto field : fieldTypes) {
+    // Create flattened block arguments.
+    auto type = getPortType(field);
+    auto newValue = addArgument(type, arg.getArgNumber(), field.suffix);
+
+    // Map the flattened suffix for the original argument to the new value.
+    setBundleLowering(arg, field.suffix, newValue);
+  }
+}
+
+// Lower any other arguments by copying them to keep the relative order.
+void FIRRTLTypesLowering::lowerArg(BlockArgument arg, Type type) {
+  auto newValue = addArgument(type, arg.getArgNumber());
+  arg.replaceAllUsesWith(newValue);
+}
+
+//===----------------------------------------------------------------------===//
+// Lowering operations.
+//===----------------------------------------------------------------------===//
+
+// Lower instance operations in the same way as module block arguments. Bundles
+// are flattened, and other arguments are copied to keep the relative order. By
+// ensuring both lowerings are the same, we can process every module in the
+// circuit in parallel, and every instance will have the correct ports.
+LogicalResult FIRRTLTypesLowering::visitDecl(InstanceOp op) {
+  auto originalType = op.result().getType().dyn_cast<BundleType>();
+  if (!originalType) {
+    op.emitError("instance result was not bundle type");
```

Comment:
I think we actually know this will be a bundle type statically. The documentation for the `InstanceOp` says "The result is always a bundle".

Reply:
Would it make sense to add a `BundleType getResultType()` function to `InstanceOp`, to convey the invariant that this is always a `BundleType` (and to assert)?

Reply:
Yeah, I think that would be nicer. I think we could also define the results to be a single bundle explicitly in ODS. As in, changing this
```
  let results = (outs FIRRTLType:$result);                                                                                                                                                                         
```
to 
```
  let results = (outs BundleType:$result);                                                                                                                                                                         
```
I'll see what works best.